reps/2023-08-18-serve-java-dag-api.md

```diff
@@ -0,0 +1,225 @@
+  }
+
+  public static class ModelA {
+    public String call(String msg) {
+      return msg;
+    }
+  }
+
+  public static class ModelB {
+    public String call(String msg) {
+      return msg;
+    }
+  }
+
+  public static void main(String[] args) {
+    Application modelA = Serve.deployment().setDeploymentDef(ModelA.class.getName()).bind();
+    Application modelB = Serve.deployment().setDeploymentDef(ModelB.class.getName()).bind();
+
+    Application driver =
+        Serve.deployment().setDeploymentDef(Driver.class.getName()).bind(modelA, modelB);
+    Serve.run(driver);
+  }
+}
+
+```
+In this example, the modelA and modelB are defined as two Deployments, and the driver is instantiated with the corresponding `DeploymentHandle` of these two Deployments. When `call` is executed, both models are invoked. Additionally, it is evident that `DeploymentHandle.remote` returns `DeploymentResponse` instead of `ObjectRef`. The result can be accessed through `DeploymentResponse.result`.
+
+### Deploying Deployments of the Other Languages through Python API
+In another REP ([Add Cpp Deployment in Ray Serve](https://github.com/ray-project/enhancements/pull/34)), it is mentioned how to deploy C++ deployments through Python. Deploying Java deployments through Python is similar. Since Java and C++ do not have the decorator mechanism like Python, a straightforward way is to directly use the `serve.deployment` API (with the addition of a new `language` parameter):
+
+```python
+deployment = serve.deployment('io.ray.serve.ExampleDeployment', name='my_deployment', language=JAVA)
+
+```
+### Deploying through the Config File
+Using a config file, we can deploy the Serve application through the CLI. For Java deployment, we can also describe it in a Python file and generate the corresponding deployment configuration in the config file. Let's take the example of the text.py file:
+```python
+from ray import serve
+from ray.serve.generated.serve_pb2 import JAVA
+
+
+@serve.deployment
+class Hello:
+    def __call__(self) -> str:
+        return "Hello"
+
+
+world_java = serve.deployment('io.ray.serve.World', language=JAVA)
+
+
+@serve.deployment
+class Ingress:
+    def __init__(self, hello_handle, world_handle):
+        self._hello_handle = hello_handle.options(
+            use_new_handle_api=True,
+        )
+        self._world_handle = world_handle.options(
+            use_new_handle_api=True,
+        )
+
+    async def __call__(self) -> str:
+        hello_response = self._hello_handle.remote()
+        world_response = self._world_handle.remote()
+        return (await hello_response) + (await world_response)
+
+
+hello = Hello.bind()
+world = world_java.bind()
+
+app = Ingress.bind(hello, world)
+
+```
+In this code, we define the Java Deployment `World` using the Python API and then bind it with a Python Deployment `Hello` into the `Ingress` to form an app. By using the `serve build` command, the Serve config file can be generated.
```

Comment:
If we only wanted to deploy `"io.ray.serve.World"`, would we still need to define this Python file?

Ideally we would be able to provide a java import path directly under `applications`. For example:

```
applications:
- name: python_app
   route_prefix: /python
   import_path: python_file:app
- name: java_app
   route_prefix: /java
   import_path: io.ray.serve.World # This needs to be some kind of similar "app builder" like the Python version!
```

Reply:
In python we have support for defining an "app builder" function which is imported and run with some provided args, then returns the built app object. Docs:
https://docs.ray.io/en/latest/serve/advanced-guides/app-builder-guide.html?highlight=builder#defining-an-application-builder

I think we could mirror this directly for Java.

Reply:
This suggestion is excellent, and I hadn't considered this aspect thoroughly before. I have now added the necessary supplements to address this.